Fabric.Authorization.API/Modules/UsersModule.cs

```diff
@@ -123,20 +123,64 @@ public class UsersModule : FabricModule<User>
                     HttpStatusCode.BadRequest);
             }
 
+            var requestErrors = new List<string>();
+
+            var permissionsWithMissingIds = permissions.Where(p => !p.Id.HasValue).ToList();
+            var permissionsWithInvalidActions = permissions.Where(p => p.PermissionAction != PermissionAction.Allow
+                                                                       && p.PermissionAction != PermissionAction.Deny)
+                .ToList();
+
+            if (permissionsWithMissingIds.Any())
+            {
+                requestErrors.AddRange(permissionsWithMissingIds.Select(p => $"{p.Name} is missing its id property."));
+            }
+
+            if (permissionsWithInvalidActions.Any())
+            {
+                requestErrors.AddRange(permissionsWithInvalidActions.Select(p => $"{p.Name} {p.Id} does not have a valid permissionAction."));
+            }
+
+            if (requestErrors.Any())
+            {
+                return CreateFailureResponse(requestErrors, HttpStatusCode.BadRequest);
+            }
+
+            var allowedPermissions = new List<Permission>();
+            var deniedPermissions = new List<Permission>();
+
             foreach (var perm in permissions)
             {
                 await CheckAccess(_clientService, perm.Grain, perm.SecurableItem, AuthorizationManageClientsClaim);
+
+                // should never hit this with validation above, but the GetPermission call below
+                // complains with a possible InvalidOperationException
+                if (!perm.Id.HasValue)
+                {
+                    continue;
+                }
+
+                var permission = await _permissionService.GetPermission(perm.Id.Value);

```

Comment:
Instead of querying for each permission in a loop, is it possible to query all the permissions in a single query so we don't have to potentially execute lots of queries here?